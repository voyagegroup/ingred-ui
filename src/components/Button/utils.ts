import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
