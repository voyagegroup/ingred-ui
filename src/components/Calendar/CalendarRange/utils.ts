import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
