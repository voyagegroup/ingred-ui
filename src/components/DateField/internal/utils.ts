export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
