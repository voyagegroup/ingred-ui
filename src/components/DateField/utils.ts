import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
