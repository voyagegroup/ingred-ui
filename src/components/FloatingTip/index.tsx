export { default } from "./FloatingTip";
export type { FloatingTipProps, FloatingTipCloseReason } from "./FloatingTip";
