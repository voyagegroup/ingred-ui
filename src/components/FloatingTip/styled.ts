import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  align-items: flex-start;
`;

export const ContentWrapper = styled.div`
  padding: ${({ theme }) => theme.spacing * 2}px;
`;

export const IconWrapper = styled.div`
  cursor: pointer;
  padding: ${({ theme }) => theme.spacing}px;
`;
