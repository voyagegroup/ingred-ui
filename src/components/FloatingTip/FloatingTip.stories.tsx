import * as React from "react";
import styled from "styled-components";
import FloatingTip from ".";
import Icon from "../Icon";
import { FloatingTipProps } from "./FloatingTip";

const Container = styled.div`
  padding: ${({ theme }) => theme.spacing * 3}px;
  background-color: ${({ theme }) => theme.palette.background.dark};
`;

const RowContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 300px;
  margin: ${({ theme }) => theme.spacing * 3}px;
  padding: ${({ theme }) => theme.spacing * 3}px;
  background-color: ${({ theme }) => theme.palette.background.default};
`;

const IconWrapper = styled.div`
  cursor: pointer;
`;

export default {
  title: "Components/Data Display/FloatingTip",
  component: FloatingTip,
};

export const Sample: React.FunctionComponent<FloatingTipProps> = (args) => {
  const [iconWrapperElement, setIconWrapperElement] =
    React.useState<HTMLDivElement | null>(null);
  const [isOpen, setIsOpen] = React.useState<boolean>(false);
  const handleIsOpen = (isOpen: boolean) => () => {
    setIsOpen(isOpen);
  };
  return (
    <Container>
      <p>Source code is written in &rdquo;Story&rdquo; Tab at footer.</p>
      <RowContainer>
        <IconWrapper
          ref={setIconWrapperElement}
          onClick={handleIsOpen(!isOpen)}
        >
          <Icon name="question" type="fill" />
        </IconWrapper>
        <FloatingTip
          {...args}
          baseElement={iconWrapperElement}
          isOpen={isOpen || !!args.isOpen}
          onClose={handleIsOpen(false)}
        >
          <p>Some message.</p>
        </FloatingTip>
      </RowContainer>
    </Container>
  );
};
