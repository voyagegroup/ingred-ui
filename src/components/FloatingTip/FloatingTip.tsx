import * as PopperJS from "@popperjs/core";
import * as React from "react";
import { useTheme } from "../../themes";
import Icon from "../Icon";
import { ModalCloseReason } from "../Modal";
import Popover, { PopoverProps } from "../Popover";
import * as Styled from "./styled";

export type FloatingTipCloseReason = "clickCloseIcon";

export type FloatingTipProps = {
  /**
   * That becomes position reference of this component.
   */
  baseElement: HTMLElement | null;
  /**
   * Define priority of position. Please check [this](https://popper.js.org/docs/v2/modifiers/flip/#fallbackplacements).
   */
  positionPriority?: PopperJS.Placement[];
  offset?: [number, number];
  isOpen: boolean;
  children: React.ReactNode;
  onClose?: (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>,
    reason: ModalCloseReason | FloatingTipCloseReason,
  ) => void;
  /**
   * props of [Popover](/?path=/docs/components-utils-popover)
   */
  popoverProps?: Partial<PopoverProps>;
};

const FloatingTip = React.forwardRef<HTMLDivElement, FloatingTipProps>(
  (
    {
      baseElement,
      positionPriority = ["right-start"],
      offset = [0, 10],
      isOpen,
      onClose,
      popoverProps,
      children,
    },
    ref,
  ) => {
    const theme = useTheme();

    const handleClickCloseIcon = (
      event: React.MouseEvent<HTMLDivElement, MouseEvent>,
    ) => {
      if (onClose) onClose(event, "clickCloseIcon");
    };

    return (
      <Popover
        isOpen={isOpen}
        baseElement={baseElement}
        positionPriority={positionPriority}
        offset={offset}
        onClose={onClose}
        {...popoverProps}
      >
        <Styled.Container ref={ref}>
          <Styled.ContentWrapper>{children}</Styled.ContentWrapper>
          <Styled.IconWrapper onClick={handleClickCloseIcon}>
            <Icon name="close" color={theme.palette.black} />
          </Styled.IconWrapper>
        </Styled.Container>
      </Popover>
    );
  },
);

export default FloatingTip;
