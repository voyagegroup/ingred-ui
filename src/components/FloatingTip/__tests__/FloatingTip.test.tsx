import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import FloatingTip from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

jest.mock("react-dom", () => {
  const original = jest.requireActual("react-dom");
  return {
    ...original,
    createPortal: (node: any) => node,
  };
});

describe("FloatingTip component testing", () => {
  afterEach(cleanup);

  test("FloatingTip", () => {
    const { asFragment } = renderWithThemeProvider(
      <FloatingTip baseElement={null} isOpen={true} onClose={() => {}}>
        <div>hoge</div>
      </FloatingTip>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
