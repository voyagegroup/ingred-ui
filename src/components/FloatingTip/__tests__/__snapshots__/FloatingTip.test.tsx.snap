// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingTip component testing FloatingTip 1`] = `
<DocumentFragment>
  <div
    class="sc-eCImPb gEZCVL"
  >
    <div
      class="sc-gKclnd frGHeV fade-transition-appear fade-transition-appear-active"
    >
      <div
        class="sc-jRQBWg iRPcoZ"
      />
    </div>
    <div
      class="sc-hKwDye kvypRT"
      style="position: absolute; left: 0px; top: 0px;"
    >
      <div
        class="sc-bdvvtL tDhDe"
      >
        <div
          class="sc-gsDKAQ gpNykI"
        >
          <div>
            hoge
          </div>
        </div>
        <div
          class="sc-dkPtRN bXeSQl"
        >
          <div
            class="sc-iCfMLu bdvlpC"
            size="18"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#041C33"
                height="23.141"
                rx="2.571"
                transform="translate(18.364 2) rotate(45)"
                width="5.142"
              />
              <rect
                fill="#041C33"
                height="23.141"
                rx="2.571"
                transform="translate(2 5.636) rotate(-45)"
                width="5.143"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
